### PR TITLE
[LibGit2] Remove excess patch and update libgit2 to latest mbedtls dependency

### PR DESCRIPTION
**Pull request by jeremiahpslewis**
- Remove upstreamed patches (this needs a careful review, please!)
- Update libgit2 version
- Update mbedtls version
- Update libgit2-hostkey patch (this needs a careful review, please!)

### CODE REVIEW

**Comment by nalimilan** [comment]
Builds failed because libgit2-hostkey.patch doesn't apply cleanly.

**Comment by jeremiahpslewis** [comment]
@nalimilan Good new, bad news...

Good news: fixed the patch so that it runs through.
Bad news: first time I have ever touched C code, so absolutely no idea what it does...

File: L/LibGit2/bundled/patches/libgit2-hostkey.patch [hunk 1]
```diff
@@ -1,24 +1,5 @@
...
- 	 * have the SHA-256 hash of the hostkey.
- 	 */
- 	unsigned char hash_sha256[32];
-+
-+	/**
-+	 * Hostkey itself.
-+	 */
-+	int hostkey_type;
-+	size_t hostkey_len;
-+	unsigned char hostkey[1024];
```
Comment (nalimilan) [inline L/LibGit2/bundled/patches/libgit2-hostkey.patch:15]: Mmm, why did you remove this? It sounds like an essential part of the patch.

Comment (jeremiahpslewis) [reply L/LibGit2/bundled/patches/libgit2-hostkey.patch:15]: Because it seems to be handled by upstream code now and looked redundant.

Comment (nalimilan) [reply L/LibGit2/bundled/patches/libgit2-hostkey.patch:15]: Great, indeed I see it at https://github.com/libgit2/libgit2/blob/v1.2.0/include/git2/cert.h#L108-L151.

File: L/LibGit2/bundled/patches/libgit2-agent-nonfatal.patch [hunk 2]
```diff
@@ -1,25 +0,0 @@
...
-@@ -296,8 +296,10 @@ static int ssh_agent_auth(LIBSSH2_SESSION *session, git_cred_ssh_key *c) {
- 
- 	rc = libssh2_agent_connect(agent);
- 
--	if (rc != LIBSSH2_ERROR_NONE)
-+	if (rc != LIBSSH2_ERROR_NONE) {
-+		rc = LIBSSH2_ERROR_AUTHENTICATION_FAILED;
- 		goto shutdown;
-+	}
-
```
Comment (nalimilan) [inline L/LibGit2/bundled/patches/libgit2-agent-nonfatal.patch:23]: This patch is still needed AFAICT: https://github.com/libgit2/libgit2/blob/v1.2.0/src/transports/ssh.c#L304-L309
Also the upstream issue isn't closed: https://github.com/libgit2/libgit2/issues/3866

Comment (jeremiahpslewis) [reply L/LibGit2/bundled/patches/libgit2-agent-nonfatal.patch:23]: Dangerous on my part to pretend to know what is going on in C code, but why wouldn't the same behavior be captured here:

https://github.com/libgit2/libgit2/blob/v1.2.0/src/transports/ssh.c#L314:L332

that's why I thought it might be redundant...

Comment (jeremiahpslewis) [reply L/LibGit2/bundled/patches/libgit2-agent-nonfatal.patch:23]: Fair enough. Do you know whether we can get someone to look at this patch and see whether it might be irrelevant (original issue is from 5 years ago)? Seems preferable to dragging this forward until the end of time?

Comment (jeremiahpslewis) [reply L/LibGit2/bundled/patches/libgit2-agent-nonfatal.patch:23]: Especially as according to this comment (from a separate, but related issue) from 2 years ago (3 years after the issue was opened), there was a significant refactor of the error handling code...

> We've significantly refactored the network code and w/r/t authentication; you should now only get a second request for credentials if the first ones were actually refused. (Credentials persist through redirects and other replays.) Closing as obsolete.

https://github.com/libgit2/libgit2/issues/3761#issuecomment-500854417

Comment (nalimilan) [reply L/LibGit2/bundled/patches/libgit2-agent-nonfatal.patch:23]: Hard to tell given how old is this patch. Can you confirm whether all tests pass without it (when applying the patches to a clean Julia git clone and building with `USE_BINARYBUILDER_LIBGIT2=0`.

@wildart @Keno Do you think https://github.com/JuliaLang/julia/pull/17459 is fixed upstream now? Is it enough to run the testsuite to check that?

Comment (mkitti) [reply L/LibGit2/bundled/patches/libgit2-agent-nonfatal.patch:23]: I'm just noting that the conda-forge feedstock for Julia currently uses generic libgit2 and mbedtls (e.g. `USE_SYSTEM_LIBGIT2=1 USE_SYSTEM_MBEDTLS=1`) on Linux only. There seem to be more complex issues on macOS.
https://github.com/conda-forge/julia-feedstock/blob/421845a3ca31c24beef378b78e7fe5209e7b871b/recipe/build.sh#L25
